src/peft/tuners/delora/model.py

```diff
@@ -0,0 +1,300 @@
+# Copyright 2025-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import warnings
+from dataclasses import asdict
+from enum import Enum
+from typing import Optional
+
+import torch
+from tqdm import tqdm
+
+from peft.tuners.tuners_utils import BaseTuner, BaseTunerLayer, check_target_module_exists
+from peft.utils import (
+    TRANSFORMERS_MODELS_TO_DELORA_TARGET_MODULES_MAPPING,
+    ModulesToSaveWrapper,
+    _get_submodules,
+)
+from peft.utils.other import get_pattern_key
+
+from .config import DeloraConfig
+from .layer import DeloraLayer, DeloraLinear
+
+
+class DeloraModel(BaseTuner):
+    """
+    Creates DeLoRA model from a pretrained transformers model.
+
+    The method is described in detail in [TODO].
+
+    Args:
+        model ([`torch.nn.Module`]): The model to be adapted.
+        config ([`DeloraConfig`]): The configuration of the DeLoRA model.
+        adapter_name (`str`): The name of the adapter, defaults to `"default"`.
+
+    Returns:
+        `torch.nn.Module`: The DeLoRA model.
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`DeloraConfig`]): The configuration of the DeLoRA model.
+    """
+
+    prefix: str = "delora_"
+
+    def _check_new_adapter_config(self, config: DeloraConfig) -> None:
+        """
+        A helper method to check the config when a new adapter is being added.
+
+        Raise a ValueError if there is something wrong with the config or if it conflicts with existing adapters.
+
+        """
+        # TODO: there should be a check if any of the existing adapters actually has bias != "none", or else the check
+        # does not fully correspond to the error message.
+        if (len(self.peft_config) > 1) and (config.bias != "none"):
+            raise ValueError(
+                f"{self.__class__.__name__} supports only 1 adapter with bias. When using multiple adapters, "
+                "set bias to 'none' for all adapters."
+            )
+        # if there are multiple adapters, residual init cannot be used
+        if (len(self.peft_config) > 1) and (config.use_residual_init):
+            raise ValueError(
+                f"{self.__class__.__name__} supports only 1 adapter with residual init. When using multiple adapters, "
+                "set use_residual_init to False for all adapters."
+            )
+
+    @staticmethod
```

Comment:
We recently had a PR (#2771) that allows us to greatly reduce the number of methods required to implement here. You can simply remove the following methods:

- `_check_target_module_exists`
- `_mark_only_adapters_as_trainable`
- `__getattr__`
- `get_peft_config_as_dict`
- `_set_adapter_layers`
- `enable_adapter_layers`
- `disable_adapter_layers`
- `set_adapter`
- `_prepare_adapter_config`
- `_unload_and_optionally_merge`
- `merge_and_unload`
- `unload`

On the flip side, you need to add:

```python
    tuner_layer_cls = DeloraLayer
    target_module_mapping = TRANSFORMERS_MODELS_TO_DELORA_TARGET_MODULES_MAPPING
```

as class attributes, i.e. directly below `prefix: str = "delora_"`.

For this to work, you also need to merge with/rebase on the latest main branch.